### PR TITLE
Revert "Merge pull request #6903 from mavasani/MakeAnalyzerExecutorAsync"

### PR DESCRIPTION
**Pull request by KevinH-MS**
This reverts commit 268e7d3804812aa7d8134c1704f07ea8e42f969c, reversing
changes made to 98790c02be4adba6cd1f785a8932f7dc83a862a2.

Conflicts:
	src/Compilers/Core/Portable/DiagnosticAnalyzer/AnalyzerDriver.cs

This change seems to be largely responsible for extra allocations and VM usage during "Start Debugging" scenarios in RPS tests.  I'm going to revert this change until we can assess the reason for that.

### CODE REVIEW

**Comment by davkean** [comment]
The failures in eta look like real failures due to the revert.

**Comment by KevinH-MS** [comment]
@davkean, yeah, I'm fixing that and will push an update shortly.

**Comment by davkean** [comment]
Thanks for chasing this down though. Appreciate it.

**Comment by davkean** [comment]
tag @jmarolf @mavasani

File: src/Compilers/CSharp/Test/Semantic/Diagnostics/DiagnosticAnalyzerTests.cs [hunk 1]
```diff
@@ -1134,77 +1134,5 @@ class D
...
-
-            // NOTE: We set the concurrentBuild option to true after creating the compilation as CreateCompilationWithMscorlib
-            //       always sets concurrentBuild to false if debugger is attached, even if we had passed options with concurrentBuild = true to that API.
-            //       We want the tests using GetCompilationWithConcurrentBuildEnabled to have identical behavior with and without debugger being attached.
-            var options = compilation.Options.WithConcurrentBuild(true);
-            return compilation.WithOptions(options);
-        }
-
-        [Fact, WorkItem(6737, "https://github.com/dotnet/roslyn/issues/6737")]
-        public void TestNonConcurrentAnalyzer()
```
Comment (mavasani) [inline src/Compilers/CSharp/Test/Semantic/Diagnostics/DiagnosticAnalyzerTests.cs]: These tests were added for a separate change, please don't remove them.

File: src/Compilers/Core/Portable/DiagnosticAnalyzer/AnalyzerDriver.cs [hunk 2]
```diff
@@ -164,6 +165,9 @@ private void Initialize(AnalyzerExecutor analyzerExecutor, DiagnosticQueue diagn
                 newOnAnalyzerException = (ex, analyzer, diagnostic) => addDiagnostic(diagnostic);
             }
 
+            // Assume all analyzers are non-thread safe.
```
Comment (mavasani) [inline src/Compilers/Core/Portable/DiagnosticAnalyzer/AnalyzerDriver.cs]: Even this was done for a separate change, please remove unused locals singleThreadedAnalyzerToGateMap and getAnalyzerGate.

File: src/Test/Utilities/Desktop/CommonDiagnosticAnalyzers.cs [hunk 3]
```diff
@@ -480,120 +480,5 @@ private void OnCompilation(CompilationAnalysisContext context)
...
             }
         }
-
-        /// <summary>
-        /// This analyzer is intended to be used only when concurrent execution is enabled for analyzers.
-        /// This analyzer will deadlock if the driver runs analyzers on a single thread OR takes a lock around callbacks into this analyzer to prevent concurrent analyzer execution
-        /// Former indicates a bug in the test using this analyzer and the latter indicates a bug in the analyzer driver.
-        /// </summary>
-        [DiagnosticAnalyzer(LanguageNames.CSharp, LanguageNames.VisualBasic)]
-        public class ConcurrentAnalyzer : DiagnosticAnalyzer
```
Comment (mavasani) [inline src/Test/Utilities/Desktop/CommonDiagnosticAnalyzers.cs]: Please revert the changes in this file, they were added for a separate change.

File: src/Compilers/Core/Portable/DiagnosticAnalyzer/AnalyzerDriver.cs (hunk 2 above)
Comment (KevinH-MS) [inline src/Compilers/Core/Portable/DiagnosticAnalyzer/AnalyzerDriver.cs]: They were touched by the original PR, so they got reset to their original state.  I didn't notice they were unused.  Will remove.

**Comment by mavasani** [comment]
@KevinH-MS Do we know if the extra allocations are coming from here: http://source.roslyn.io/#Microsoft.CodeAnalysis/InternalUtilities/SemaphoreSlimExtensions.cs,20? The original change being reverted just replaces a bunch of explicit lock statements on gate objects with DisposableWaitAsync on SemaphoreSlim, and it seems we are instantiating a new SemaphoreDisposer for every such invocation. If the revert is not blocking any further perf validations, I can attempt a fix to address this tomorrow and validate perf. Otherwise, feel free to revert it and I'll submit a new PR to restore the change with the fix after doing perf validations.

File: src/Test/Utilities/Desktop/CommonDiagnosticAnalyzers.cs (hunk 3 above)
Comment (KevinH-MS) [inline src/Test/Utilities/Desktop/CommonDiagnosticAnalyzers.cs]: It required the semaphoreslim extensions reverted with this PR.  I can put those back and see if it compiles.

File: src/Test/Utilities/Desktop/CommonDiagnosticAnalyzers.cs [hunk 4]
```diff
@@ -480,120 +480,5 @@ private void OnCompilation(CompilationAnalysisContext context)
...
-                "Analyzer driver made concurrent action callbacks, when analyzer didn't register for concurrent execution",
-                "Category",
-                DiagnosticSeverity.Warning,
-                true);
-            private const int registeredActionCounts = 1000;
-
-            public override ImmutableArray<DiagnosticDescriptor> SupportedDiagnostics => ImmutableArray.Create(Descriptor);
-            public override void Initialize(AnalysisContext context)
-            {
-                SemaphoreSlim gate = new SemaphoreSlim(initialCount: registeredActionCounts);
```
Comment (mavasani) [inline src/Test/Utilities/Desktop/CommonDiagnosticAnalyzers.cs]: Ah, the new test I added also used the SemaphoreSlim utilities. Can you retains these tests and also the move of SemaphoreSlim utilities to CodeAnalysis? It will have test only consumers for now, but I hope to restore the original PR with the fix soon.

Comment (mavasani) [inline src/Test/Utilities/Desktop/CommonDiagnosticAnalyzers.cs]: If it is too much trouble, feel free to keep what you have in your current PR - I will restore even these tests with my new PR.

**Comment by KevinH-MS** [comment]
@mavasani I've updated the PR per your comments.

Unfortunately, I'm not sure what's causing the extra allocations, but I've e-mailed you links to the profiles (before/after).

I apologize for the heavy-handed approach, but we really need to get things back to a "good" state asap (we haven't integrated in over a month now), and there are several causes of regressions that I need to isolate.

**Comment by mavasani** [comment]
Thanks :+1:
